periodic_builds/tasks.py

```diff
@@ -0,0 +1,76 @@
+# -*- coding: utf-8 -*-
+#
+# OpenCraft -- tools to aid developing and hosting free software projects
+# Copyright (C) 2019 OpenCraft <contact@opencraft.com>
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+"""
+Worker tasks for periodically building new open edx appservers at intervals.
+"""
+
+# Imports #####################################################################
+
+import logging
+import datetime
+
+from huey.api import crontab
+from huey.contrib.djhuey import db_periodic_task
+
+from instance.models.openedx_instance import OpenEdXInstance
+from instance.models.appserver import Status
+from instance.tasks import spawn_appserver
+
+
+# Logging #####################################################################
+
+logger = logging.getLogger(__name__)
+
+
+# Tasks #######################################################################
+
+# TODO: is every 30 minutes ok here?
+@db_periodic_task(crontab(minute="*/30"))
+def launch_periodic_builds():
+    """
+    Automatically deploy new servers for all Open edX instances configured for periodic builds.
+    """
+
+    # get all instances configured for periodic builds
+    instances = OpenEdXInstance.objects.filter(periodic_builds_enabled=True)
+
+    now = datetime.datetime.now(tz=datetime.timezone.utc)
+    for instance in instances:
+        # if the instance is due for a re-build, then we spawn a new appserver
+        latest_appserver = instance.appserver_set.latest("created")
```

Comment:
Try to resolve appserver_set in tests when I'm testing locally just results in the entire function silently returning. Don't know why.

Reply:
fixed: turns out the huey decorator suppresses errors. removed the decorator for debugging, and then re-added it.